build.assets/robotest/Makefile

```diff
@@ -0,0 +1,326 @@
+ROBOTEST_TEST_CONFIG = $(TOP)/pr_config.sh
+# ROBOTEST_UPGRADE_VERSIONS is a list of gravity semvers that will be
+# used in upgrade testing.
+ROBOTEST_UPGRADE_VERSIONS = $(shell ./scrape_upgrade_from_versions.sh $(ROBOTEST_TEST_CONFIG))
+# ROBOTEST_UPGRADE_CACHE_IMAGES is a list of potentially cached robotest tarballs
+# to be used in upgrade testing.
+ROBOTEST_UPGRADE_CACHE_IMAGES = $(addsuffix .tar, $(addprefix $(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-, $(ROBOTEST_UPGRADE_VERSIONS)))
+# ROBOTEST_TELE_BINARIES is the list of build specific tele binaries needed to
+# build all robotest upgrade images.
+ROBOTEST_TELE_BINARIES = $(addprefix $(ROBOTEST_TELE_CACHE)/tele-, $(ROBOTEST_UPGRADE_VERSIONS))
+# ROBOTEST_UPGRADE_IMAGES is the list of build specific upgrade images, suitable
+# for mounting into the robotest docker container. These are not shared stated.
+ROBOTEST_UPGRADE_IMAGES = $(addsuffix .tar, $(addprefix $(ROBOTEST_IMAGEDIR)/robotest-, $(ROBOTEST_UPGRADE_VERSIONS)))
+# ROBOTEST_IMAGES is all build specific cluster images/targets needed to run robotest.
+ROBOTEST_IMAGES = $(ROBOTEST_IMAGE) $(ROBOTEST_OPSCENTER_IMAGE) $(ROBOTEST_UPGRADE_IMAGES)
+
+# Robotest run configuration.
+# These are only used in the run target.
+
+ROBOTEST_GRAVITY_BINARY ?= $(GRAVITY_OUT)
+ROBOTEST_SSH_KEY ?= $(HOME)/.ssh/id_rsa
+ROBOTEST_SSH_PUB ?= $(HOME)/.ssh/id_rsa.pub
+ROBOTEST_GOOGLE_APPLICATION_CREDENTIALS ?= $(ROOT)/robotest-gcp-credentials.json
+
+.PHONY: help
+# kudos to https://gist.github.com/prwhite/8168133 for inspiration
+help: ## Show this message.
+	@echo 'Usage: make [options] [target] ...'
+	@echo
+	@echo 'Options: run `make --help` for options'
+	@echo
+	@echo 'Targets:'
+	@egrep '^(.+)\:\ ##\ (.+)' ${MAKEFILE_LIST} | column -t -c 2 -s ':#' | sort | sed 's/^/  /'
+
+.PHONY: clean
+clean: ## Remove build artifacts.
+	rm -rf $(ROBOTEST_BUILDDIR)
+
+.PHONY: clean-caches
+clean-caches: ## Remove cached tele binaries, gravity packages, and robotest upgrade images.
+	rm -rf $(ROBOTEST_CACHES) $(ROBOTEST_CACHE_TMP)
+
+$(ROBOTEST_BUILDDIR):
+	mkdir -p $(ROBOTEST_BUILDDIR)
+
+$(ROBOTEST_UPGRADE_IMAGE_CACHE):
+	mkdir -p $(ROBOTEST_UPGRADE_IMAGE_CACHE)
+
+$(ROBOTEST_IMAGEDIR):
+	mkdir -p $(ROBOTEST_IMAGEDIR)
+
+$(ROBOTEST_TELE_CACHE):
+	mkdir -p $(ROBOTEST_TELE_CACHE)
+
+$(ROBOTEST_BUILD_TMP):
+	mkdir -p $(ROBOTEST_BUILD_TMP)
+
+$(ROBOTEST_CACHE_TMP):
+	mkdir -p $(ROBOTEST_CACHE_TMP)
+
+.PHONY:docker-image
+docker-image: ## Create robotest docker image with the current user/group baked in.
+docker-image: $(ROBOTEST_DOCKER_IMAGE_IIDFILE)
+
+$(ROBOTEST_DOCKER_IMAGE_IIDFILE): Dockerfile Makefile
+	docker build \
+		--build-arg BASE=$(ROBOTEST_DOCKER_UPSTREAM_IMAGE) \
+		--build-arg UID=$(shell id -u) \
+		--build-arg GID=$(shell id -g) \
+		--iidfile $(ROBOTEST_DOCKER_IMAGE_IIDFILE) \
+		--tag $(ROBOTEST_DOCKER_IMAGE) .
+
+.PHONY: images
+images: ## Create all images necessary to run robotest CI tests.
+images: image-robotest images-upgrade image-opscenter
```

Comment:
Consider doing this as a `$(MAKE) -j targets` to parallelize the main image and opscenter image work with the already parallel upgrade images -- alternatively, this may may the logs very difficult to follow.  If kept this way, leave a comment as to why.

Reply:
Parallelism won't work with a shared cache, as discussed in ecc6916.